packages/react/src/components/GroupedList/GroupedListV2.base.tsx

```diff
@@ -0,0 +1,507 @@
+import * as React from 'react';
+import {
+  initializeComponentRef,
+  classNamesFunction,
+  KeyCodes,
+  getRTLSafeKeyCode,
+  css,
+  getId,
+  EventGroup,
+} from '../../Utilities';
+import { List, ScrollToMode, IListProps } from '../../List';
+import { SelectionMode, SELECTION_CHANGE } from '../../Selection';
+import { FocusZone, FocusZoneDirection } from '../../FocusZone';
+import type { IProcessedStyleSet } from '../../Styling';
+import type {
+  IGroupedList,
+  IGroupedListProps,
+  IGroup,
+  IGroupRenderProps,
+  IGroupedListStyleProps,
+  IGroupedListStyles,
+} from './GroupedList.types';
+import { GroupHeader } from './GroupHeader';
+import { GroupShowAll } from './GroupShowAll';
+import { GroupFooter } from './GroupFooter';
+import type { IGroupHeaderProps } from './GroupHeader';
+import type { IGroupShowAllProps } from './GroupShowAll.styles';
+import type { IGroupFooterProps } from './GroupFooter.types';
+import { useMount, useUnmount } from '@fluentui/react-hooks';
+
+export interface IGroupedListV2State {
+  selectionMode?: IGroupedListProps['selectionMode'];
+  compact?: IGroupedListProps['compact'];
+  groups?: IGroup[];
+  items?: IGroupedListProps['items'];
+  listProps?: IGroupedListProps['listProps'];
+  version: {};
+  groupExpandedVersion: {};
+}
+
+export interface IGroupedListV2Props extends IGroupedListProps {
+  listRef: React.Ref<List>;
+  version: {};
+  groupExpandedVersion: {};
+}
+
+type IGroupedItem = {
```

Comment:
Fun typing madness:
```ts
type IItemGroupedItem = {
  type: 'item';
  item: IItem;
  itemIndex: number;
};
type IShowAllGroupedItem = {
  type: 'showAll';
  group: IGroup;
  groupId: string;
};

 type IFooterGroupedItem {
  type: 'footer';
  group: IGroup;
  groupId: string;
}

type IHeaderGroupedItem {
  type: 'header';
  group: IGroup;
  groupId: string;
};

type IGroupedItem = IItemGroupedItem | IShowAllGroupedItem | IHeaderGroupedItem | IFooterGroupedItem;
```
If you do this, then the `switch` on `type` which ensure the type-safety of the other members.